File: app/src/main/java/life/plank/juna/zone/util/customview/LineupPlayer.java
package life.plank.juna.zone.util.customview;

import android.content.Context;
import android.support.annotation.ColorRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;

import butterknife.BindView;
import butterknife.ButterKnife;
import de.hdodenhof.circleimageview.CircleImageView;
import life.plank.juna.zone.R;
import life.plank.juna.zone.data.model.Formation;

public class LineupPlayer extends FrameLayout {

    @BindView(R.id.lineup_player_pic)
    CircleImageView lineupPlayerPic;
    @BindView(R.id.lineup_player_number)
    CircularTextView lineupPlayerNumber;
    @BindView(R.id.lineup_player_card)
    ImageView lineupPlayerCard;
    @BindView(R.id.lineup_player_goal)
    ImageView lineupPlayerGoal;
    @BindView(R.id.lineup_player_goal_count)
    CircularTextView lineupPlayerGoalCount;
    @BindView(R.id.lineup_player_substitution)
    ImageView lineupPlayerSubstitution;
    @BindView(R.id.lineup_player_name)
    TextView lineupPlayerName;

    private Formation formation;

    public LineupPlayer(@NonNull Context context) {
        this(context, null, R.color.purple);
    }

    public LineupPlayer(@NonNull Context context, Formation formation, @ColorRes int labelColor) {
        this(context, null, formation, labelColor);
        this.formation = formation;
    }

    public LineupPlayer(@NonNull Context context, @Nullable AttributeSet attrs, Formation formation, @ColorRes int labelColor) {
        this(context, attrs, 0, formation, labelColor);
        this.formation = formation;
    }

    public LineupPlayer(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr, Formation formation, @ColorRes int labelColor) {
        this(context, attrs, defStyleAttr, 0, formation, labelColor);
        this.formation = formation;
    }

    public LineupPlayer(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes, Formation formation, @ColorRes int labelColor) {
        super(context, attrs, defStyleAttr, defStyleRes);
        this.formation = formation;
        init(context, labelColor);
    }

    private void init(Context context, @ColorRes int labelColor) {
        View rootView = inflate(context, R.layout.item_player_in_lineup, this);
        ButterKnife.bind(this, rootView);
        update(labelColor);
    }

    public void update(@ColorRes int labelColor) {
        if (formation != null) {
            setPlayerNumber(formation.getNumber())
//    TODO : replace with boolean values once backend filters it
                    .setPlayerCard(
                            formation.getYellowCards(),
                            formation.getRedCards(),
                            formation.getYellowRed())
                    .setSubstituted(formation.getSubstituteOut())
                    .setSolidColor(labelColor)
                    .setPlayerNumber(formation.getNumber())
                    .setGoal(formation.getGoals())
                    .setName(formation.getNickname());
            Picasso.with(getContext())
                    .load(formation.getImagePath())
                    .placeholder(R.drawable.ic_default_profile)
                    .error(R.drawable.ic_default_profile)
                    .into(lineupPlayerPic);
        }
    }

    public LineupPlayer setSolidColor(@ColorRes int labelColor) {
        lineupPlayerNumber.setSolidColor(labelColor);
        return this;
    }

    public LineupPlayer setPlayerNumber(int number) {
        this.lineupPlayerNumber.setText(String.valueOf(number));
        return this;
    }

    public LineupPlayer setPlayerCard(int yellowCard, int redCard, int yellowRed) {
        if (yellowCard == 0 && redCard == 0 && yellowRed == 0) {
            lineupPlayerCard.setVisibility(GONE);
            return this;
        }
        lineupPlayerCard.setVisibility(VISIBLE);
        if (yellowRed == 1) {
            lineupPlayerCard.setImageResource(R.drawable.yellow_red);
        } else if (redCard == 1) {
            lineupPlayerCard.setImageResource(R.drawable.red_right);
        } else if (yellowCard == 1) {
            lineupPlayerCard.setImageResource(R.drawable.yellow_right);
        }
        return this;
    }

    public LineupPlayer setGoal(int goalCount) {
        this.lineupPlayerGoal.setVisibility(goalCount > 0 ? VISIBLE : GONE);
        setGoalCount(goalCount);
        this.lineupPlayerGoalCount.setVisibility(goalCount > 1 ? VISIBLE : GONE);
        return this;
    }

    public void setGoalCount(int goalCount) {
        this.lineupPlayerGoalCount.setText(String.valueOf(goalCount));
    }

    public LineupPlayer setSubstituted(int isSubstitutedIn) {
        this.lineupPlayerSubstitution.setVisibility(isSubstitutedIn == 1 ? VISIBLE : GONE);
        return this;
    }

    public String getName() {
        return lineupPlayerName.getText().toString();
    }

    public LineupPlayer setName(String name) {
        this.lineupPlayerName.setText(name);
        return this;
    }
}

File: app/src/main/java/life/plank/juna/zone/util/customview/LineupPlayer.kt
package life.plank.juna.zone.util.customview

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout
import androidx.annotation.ColorRes
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import kotlinx.android.synthetic.main.item_player_in_lineup.view.*
import life.plank.juna.zone.R
import life.plank.juna.zone.data.model.Formation

class LineupPlayer(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int,
        private var formation: Formation?,
        @ColorRes labelColor: Int
) : FrameLayout(context, attrs, defStyleAttr, defStyleRes) {

    val name: String
        get() = lineup_player_name.text.toString()

    @JvmOverloads
    constructor(context: Context, formation: Formation? = null, @ColorRes labelColor: Int = R.color.purple) : this(context, null, formation, labelColor) {
        this.formation = formation
    }

    constructor(context: Context, attrs: AttributeSet?, formation: Formation?, @ColorRes labelColor: Int) : this(context, attrs, 0, formation, labelColor) {
        this.formation = formation
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, formation: Formation?, @ColorRes labelColor: Int) : this(context, attrs, defStyleAttr, 0, formation, labelColor) {
        this.formation = formation
    }

    init {
        init(context, labelColor)
    }

    private fun init(context: Context, @ColorRes labelColor: Int) {
        View.inflate(context, R.layout.item_player_in_lineup, this)
        update(labelColor)
    }

    fun update(@ColorRes labelColor: Int) {
        formation?.run {
            setPlayerNumber(number)
                    //    TODO : replace with boolean values once backend filters it
                    .setPlayerCard(
                            yellowCards,
                            redCards,
                            yellowRed)
                    .setSubstituted(substituteOut)
                    .setSolidColor(labelColor)
                    .setPlayerNumber(number)
                    .setGoal(goals)
                    .setName(nickname)
            Glide.with(this@LineupPlayer)
                    .load(imagePath)
                    .apply(RequestOptions.placeholderOf(R.drawable.ic_default_profile)
                            .error(R.drawable.ic_default_profile))
                    .into(lineup_player_pic)
        }
    }

    private fun setSolidColor(@ColorRes labelColor: Int): LineupPlayer {
        lineup_player_number.solidColor = labelColor
        return this
    }

    private fun setPlayerNumber(number: Int): LineupPlayer {
        this.lineup_player_number.text = number.toString()
        return this
    }

    private fun setPlayerCard(yellowCard: Int, redCard: Int, yellowRed: Int): LineupPlayer {
        if (yellowCard == 0 && redCard == 0 && yellowRed == 0) {
            lineup_player_card.visibility = View.GONE
            return this
        }
        lineup_player_card.visibility = View.VISIBLE
        when {
            yellowRed == 1 -> lineup_player_card.setImageResource(R.drawable.yellow_red)
            redCard == 1 -> lineup_player_card.setImageResource(R.drawable.red_right)
            yellowCard == 1 -> lineup_player_card.setImageResource(R.drawable.yellow_right)
        }
        return this
    }

    private fun setGoal(goalCount: Int): LineupPlayer {
        this.lineup_player_goal.visibility = if (goalCount > 0) View.VISIBLE else View.GONE
        setGoalCount(goalCount)
        lineup_player_goal_count.visibility = if (goalCount > 1) View.VISIBLE else View.GONE
        return this
    }

    private fun setGoalCount(goalCount: Int) {
        lineup_player_goal_count.text = goalCount.toString()
    }

    private fun setSubstituted(isSubstitutedIn: Int): LineupPlayer {
        lineup_player_substitution.visibility = if (isSubstitutedIn == 1) View.VISIBLE else View.GONE
        return this
    }

    private fun setName(name: String): LineupPlayer {
        this.lineup_player_name.text = name
        return this
    }
}